### DISCUSSION TITLE
How to ensure that value boxes will stay the same size on any screen/browser window size?

### DISCUSSION THREAD

**Discussion by udurraniAtPresage** [rstudio/bslib#926] [Q&A] [answered]
On a bigger browser window, value boxes display as:

![image](https://github.com/rstudio/bslib/assets/142413619/fe017af1-8c72-4725-9dc1-978935f4e23f)

But at a smaller width of browser, they look like this:
![image](https://github.com/rstudio/bslib/assets/142413619/5b4339f8-24f7-424c-83b9-97e8fc4553df)

Any ideas how to keep the size consistent, i.e., the second box should be of the same size as the others?

Following is my value box code. I have even tried string padding.

```r
value_box(
        title = str_pad(title, width = 15, side = "right", pad = " "), 
        value =   score,
        paste(
          "Average", average_score
        ),
        showcase = img(
          src = paste0(img_src, ".png"),
          height = "100%",
          width = "100%",
          style = "object-fit: contain; overflow: auto;"
        ),
        showcase_layout = showcase_top_right(),
        full_screen = TRUE
      )
```

**Comment by cpsievert** [comment, marked as answer]
What are you using to layout the value boxes? I'm pretty sure if you use `layout_column_wrap()` or `layout_column()` they'll all be the same height (by default)

**Comment by udurraniAtPresage** [comment]
Thanks! I was using `layout_columns` but replacing it with `layout_column_wrap()` worked.

**Reply by gadenbuie** [reply]
Did you happen to have `layout_columns(..., fillable = FALSE)`? With the default value, which is `fillable = TRUE`, `layout_columns()` does what I think you want:

![image](https://github.com/rstudio/bslib/assets/5420529/601a8a08-5105-4163-93f0-e5fca8cbea9c)

but when setting `fillable = FALSE` the heights no longer match:

![image](https://github.com/rstudio/bslib/assets/5420529/855cbd99-4670-4ed3-9f9b-d5fbebe602e5)


<details>
<summary>Example App</summary>

```r
library(shiny)
library(bslib)

ui <- page_fluid(
  layout_columns(
    fillable = FALSE,
    !!!purrr::map(1:3, \(x) value_box(
      paste("Value", x),
      x * 10,
      if (x != 2) "A little more info",
      showcase = bsicons::bs_icon(paste0(x, "-circle-fill"))
    ))
  )
)

server <- function(input, output, session) {

}

shinyApp(ui, server)
```

</details>

**Reply by udurraniAtPresage** [reply]
I was using the default value of `fillable = TRUE` with `layout_columns`. I just also tried `fillable = FALSE` and it makes no difference, i.e., the cards have unequal sizes. But `layout_column_wrap` with its default of `fillable = TRUE` ensures the equal sized value boxes for me. I am on version 0.6.1 of bslib.

**Reply by gadenbuie** [reply]
🤔 which browser and OS are you using? Is my example in the previous post a good reprex of your use case? Can you try my example to see if that reproduces the behavior? If not, it'd be helpful if you could tell me a little more about how your app is laid out.

**Reply by udurraniAtPresage** [reply]
I am on Microsoft Edge and Windows 10. Your example works for me with the same results that you showed (equal heights with `fillable = TRUE` in `layout_columns`). My app uses `page_navbar` with modules, so a modified version of your code would look something like this:

```
library(shiny)
library(bslib)

ui <- page_navbar(
  nav_panel(
    layout_columns(
      fillable = TRUE,
      tagList(
      !!!purrr::map(1:3, \(x) value_box(
        paste("Value", x),
        x * 10,
        if (x != 2) "A little more info",
        showcase = bsicons::bs_icon(paste0(x, "-circle-fill"))
      ))
      )
    )
  )
)

server <- function(input, output, session) {
  
}

shinyApp(ui, server)
```

**Reply by gadenbuie** [reply]
Thanks, that's a helpful step! I still wasn't able to reproduce the issue, though. Here's the app I used (notice that you don't need `tagList()`, the `!!!` in `!!!purrr::map(...)` splat out the returned list of items so it's like you gave each one to `layout_columns()` individually). I also added a few placeholder divs to simulate having something else in your app.

<details>
<summary>Example app</summary>

```r
library(shiny)
library(bslib)

ui <- page_navbar(
  title = "Issue 926",
  nav_panel(
    title = "Page 1",
    layout_columns(
      fillable = TRUE,
      !!!purrr::map(
        1:3,
        \(x) value_box(
          paste("Value", x),
          x * 10,
          if (x != 2) "A little more info",
          showcase = bsicons::bs_icon(paste0(x, "-circle-fill"))
        )
      )
    ),
    as_fill_item(div(class = "bg-secondary-subtle", style = css(min_height = "200px"))),
    as_fill_item(div(class = "bg-secondary-subtle", style = css(min_height = "200px")))
  )
)

server <- function(input, output, session) {

}

shinyApp(ui, server)
```

</details>


Here's what I see when I try the above example at various sizes (it's working as expected for me). Is this what you see? Is there something else we're missing in the reprex that you're using in your real-world app?


https://github.com/rstudio/bslib/assets/5420529/ac5fdd2b-d4c5-4da8-8607-eca00276b97b

**Reply by udurraniAtPresage** [reply]
This also works for me as you showed. The only other thing different in my code is that my value boxes are rendered via `renderUI` because I update the `value` argument after some calculations in the server. The ui part of the module contains `uiOutput`.  Could that be an issue?

**Reply by gadenbuie** [reply]
Yes that could be it! Are you rendering each value box individually? If so, be sure to set `uiOutput(fill = TRUE)`. Alternatively you could move the entire `layout_columns()` into the `renderUI()`. Either one should work! Let me know how that goes 😄

**Reply by udurraniAtPresage** [reply]
Yes, I am rendering each value box individually via a module. Setting `uiOutput(fill = TRUE)` works! Thank you.

**Reply by udurraniAtPresage** [reply]
Related question: `layout_columns` works great at resizing the value boxes on smaller screens. However, a longer title in a value box splits into two lines. For example, on iPad Pro, the title "Professionalism" splits to "Profession" on first line and "alism" in the second line. Is there a bslib way to automatically resize the font to keep the title in a single line?

**Reply by gadenbuie** [reply]
That's great to hear about `layout_columns()`! Unfortunately, we don't have an easy or automatic way to resize the text in the value box to match the container size. It'd be a neat feature! For now, I'd recommend trying to write some CSS to target that value box.
